File: LLM-Translations/GPT-4/codenet/Java/Go/s918182055.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	N := nextInt(scanner)
	M := nextInt(scanner)

	list := make([][]int, 100001)

	for i := 0; i < N; i++ {
		a := nextInt(scanner)
		b := nextInt(scanner)
		if list[a] == nil {
			list[a] = []int{}
		}
		list[a] = append(list[a], b)
	}

	res := 0
	pqueue := &priorityQueue{}

	for i := 1; i <= M; i++ {
		if list[i] != nil {
			pqueue.addAll(list[i])
		}
		if !pqueue.isEmpty() {
			res += pqueue.poll()
		}
	}

	fmt.Println(res)
}

type priorityQueue struct {
	data []int
}

func (pq *priorityQueue) addAll(nums []int) {
	for _, num := range nums {
		pq.push(num)
	}
}

func (pq *priorityQueue) push(num int) {
	pq.data = append(pq.data, num)
	i := len(pq.data) - 1
	for i > 0 {
		parent := (i - 1) / 2
		if pq.data[parent] >= pq.data[i] {
			break
		}
		pq.data[parent], pq.data[i] = pq.data[i], pq.data[parent]
		i = parent
	}
}

func (pq *priorityQueue) poll() int {
	res := pq.data[0]
	pq.data[0] = pq.data[len(pq.data)-1]
	pq.data = pq.data[:len(pq.data)-1]
	i := 0
	for {
		left := 2*i + 1
		right := 2*i + 2
		if left >= len(pq.data) {
			break
		}
		child := left
		if right < len(pq.data) && pq.data[right] > pq.data[left] {
			child = right
		}
		if pq.data[i] >= pq.data[child] {
			break
		}
		pq.data[i], pq.data[child] = pq.data[child], pq.data[i]
		i = child
	}
	return res
}

func (pq *priorityQueue) isEmpty() bool {
	return len(pq.data) == 0
}

func nextInt(scanner *bufio.Scanner) int {
	scanner.Scan()
	i, _ := strconv.Atoi(scanner.Text())
	return i
}

// 

File: LLM-Translations/GPT-4/codenet/Java/C/s918182055.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct LinkedList {
    int value;
    struct LinkedList *next;
} LinkedList;

typedef struct PriorityQueue {
    int size;
    int capacity;
    int *data;
} PriorityQueue;

PriorityQueue* create_pqueue(int capacity);
void add(PriorityQueue *pqueue, int value);
int poll(PriorityQueue *pqueue);
bool is_empty(PriorityQueue *pqueue);
void swap(int *a, int *b);

int main() {
    int N, M;
    scanf("%d %d", &N, &M);
    LinkedList *list[100001] = {0};

    for (int i = 0; i < N; i++) {
        int a, b;
        scanf("%d %d", &a, &b);
        LinkedList *node = (LinkedList *)malloc(sizeof(LinkedList));
        node->value = b;
        node->next = list[a];
        list[a] = node;
    }

    int res = 0;
    PriorityQueue *pqueue = create_pqueue(N);
    for (int i = 1; i <= M; i++) {
        LinkedList *cur = list[i];
        while (cur != NULL) {
            add(pqueue, cur->value);
            cur = cur->next;
        }
        if (!is_empty(pqueue)) {
            res += poll(pqueue);
        }
    }

    printf("%d\n", res);
    return 0;
}

PriorityQueue* create_pqueue(int capacity) {
    PriorityQueue *pqueue = (PriorityQueue *)malloc(sizeof(PriorityQueue));
    pqueue->size = 0;
    pqueue->capacity = capacity;
    pqueue->data = (int *)malloc(capacity * sizeof(int));
    return pqueue;
}

void add(PriorityQueue *pqueue, int value) {
    pqueue->data[pqueue->size] = value;
    int idx = pqueue->size;
    while (idx > 0 && pqueue->data[(idx - 1) / 2] < pqueue->data[idx]) {
        swap(&pqueue->data[(idx - 1) / 2], &pqueue->data[idx]);
        idx = (idx - 1) / 2;
    }
    pqueue->size++;
}

int poll(PriorityQueue *pqueue) {
    int result = pqueue->data[0];
    pqueue->data[0] = pqueue->data[--pqueue->size];

    int idx = 0;
    while (true) {
        int left = idx * 2 + 1;
        int right = idx * 2 + 2;

        if (left >= pqueue->size) break;

        int max_child = left;
        if (right < pqueue->size && pqueue->data[left] < pqueue->data[right]) {
            max_child = right;
        }

        if (pqueue->data[idx] >= pqueue->data[max_child]) break;

        swap(&pqueue->data[idx], &pqueue->data[max_child]);
        idx = max_child;
    }

    return result;
}

bool is_empty(PriorityQueue *pqueue) {
    return pqueue->size == 0;
}

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

//
